testing/pr-testing-guidelines-and-checklists.md

```diff
@@ -0,0 +1,149 @@
+
+## Problem Description
+
+We currently don't have a proper set of guidelines for our PR checks and previously we used them mostly just to produce the build-artifacts
+in some fashion. The quality checks have often been added in ad-hoc manner and then disregarded/forgotten about.
+
+## Expectations
+
+If we create a set of coherent guidelines and model our automated PR checks by them we expect
+- to catch regressions early
+- to maintain a good quality of incoming code
+
+## Terms
+
+* relevant repository - repository that could influence ship-able artifacts by it's contents in a bad way. If we really don't want to follow these guidelines, just claim that the repository is currently irrelevant for our project.
+  
+## Proposed Solution
+
+These proposals only concern relevant repositories.
+
+### Repository should have a mandatory review by other person
+
+This is easily enabled in GitHub. It mandates that at least one person gives a PR a green check-mark.
+
+There will be some decisions that we don't want to automate on purpose, for example, we will probably display code-coverage prominently,
+but we won't have any automatically enforced that would stop PR from getting merged due to low coverage.
+
+In the future, we might have optional integration tests triggered by comments or specific tags on the PR. This would be hard to enforce automatically,
+so mandating those would fall as a responsibility of the reviewer.
+
+There could be a template with a checklist delivered by a .github/PULL_REQUEST_TEMPLATE.md
+Checklist would include:
+
+* Read through the code and ask about any unclear code
+* Try to checkout and run the relevant code
+* Evaluate if the new code is sufficiently tested, potentially consulting the code-coverage
+* Evaluate if there are any follow-up tasks to be logged
+* more?
+
+### All of the automated testing should finish in reasonable time, defined as in less than 10 minutes
+
+Looking at our previous PR checking attempts, even fast checks take around 2 minutes, due to inconsistencies in time of spawning our on-demand slaves.
+
+Our longest PR builds can take around 20 minutes.
+
+Any time, there is PR check that consistently hits the 10 minute run-time, it should become an issue in our backlog to be solved.
+
+### There should be a consistent and enforced style for our code
+
+Maintaining a repository where multiple developers collaborate should be helped by ensuring consistent style.
+While this often leads to a lot of bike-shedding, and changing the style after it has been applied can be tedious,
+but the style enforcement still seems to be quite useful, especially for new collaborators.
+
+For configuration files and dynamic languages, style checking can often find basic problems that would be discovered in compiled languages with type-checker.
+
+We could experiment even with automatic style formatting, such as:
+* gofmt
+* grunt-jsbeautifier
+
+We should enforce the style for:
```

Comment:
What would be the code styles? 
There is https://github.com/fheng/best_practice, but this is not public - We should move that to aerogear proposals as well.